src/octree/node_iterator.rs

```diff
@@ -65,24 +66,32 @@ impl NodeIterator {
                 position_encoding,
                 num_points,
             },
+            point_count: 0,
         };
         Ok(NodeIterator::WithData(iter))
     }
 }
 
-impl InternalIterator for NodeIterator {
-    fn size_hint(&self) -> Option<usize> {
-        match self {
-            NodeIterator::WithData(ref iter) => Some(iter.meta.num_points as usize),
-            NodeIterator::Empty => Some(0),
-        }
+impl Iterator for NodeIterator {
+    type Item = Point;
+
+    fn size_hint(&self) -> (usize, Option<usize>) {
+        let num_points = match self {
+            NodeIterator::WithData(ref iter) => iter.meta.num_points as usize,
+            NodeIterator::Empty => 0,
+        };
+        (num_points, Some(num_points))
     }
-    fn for_each<F: FnMut(&Point)>(self, mut f: F) {
+    fn next(&mut self) -> Option<Point> {
         let mut iter = match self {
```

Comment:
I think the slowdown comes from effectively pulling everything from here until the match statement into the innermost loop. Pulling them out again would require caching `edge_length`, `min` and `num_points as usize` in the iterator. The `WithData` indirection can probably only be avoided by directly storing `NodeIteratorWithData` in the high-level iterators. Do you think any of that would be worth doing?

Reply:
`edge_length`, `min`, and `num_points` are only double or 3d doubles, which shouldn't be much overhead... For the rest I'm not sure whether this is worth the effort.